launch/launch/substitutions/anon_name.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2020 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for the anonymous name substitution."""
+
+from typing import Iterable
+from typing import List
+from typing import Text
+
+from ..frontend import expose_substitution
+from ..launch_context import LaunchContext
+from ..some_substitutions_type import SomeSubstitutionsType
+from ..substitution import Substitution
+
+
+@expose_substitution('anon')
+class AnonName(Substitution):
+    """
+    Generates an anonymous id based on name.
+
+    Name itself is a unique identifier: multiple uses of $(anon foo)
+    will create the same "anonymized" name
+    """
+
+    def __init__(self, name: SomeSubstitutionsType) -> None:
+        """Create a PythonExpression substitution."""
+        super().__init__()
+
+        from ..utilities import normalize_to_list_of_substitutions
+        self.__name = normalize_to_list_of_substitutions(name)
+
+    @classmethod
+    def parse(cls, data: Iterable[SomeSubstitutionsType]):
+        """Parse `AnonName` substitution."""
+        if len(data) != 1:
+            raise TypeError('anon substitution expects 1 argument')
+        return cls, {'name': data[0]}
+
+    @property
+    def name(self) -> List[Substitution]:
+        """Getter for name."""
+        return self.__name
+
+    def describe(self) -> Text:
+        """Return a description of this substitution as a string."""
+        return 'AnonName({})'.format(' + '.join([sub.describe() for sub in self.name]))
+
+    def perform(self, context: LaunchContext) -> Text:
+        """Perform the substitution by creating/getting the anonymous name."""
+        from ..utilities import perform_substitutions
+        name = perform_substitutions(context, self.name)
+
+        if('anon' not in context.launch_configurations):
+            context.launch_configurations['anon'] = {}
+        anon_context = context.launch_configurations['anon']
+
+        if(name in anon_context):
+            return anon_context[name]
+
+        value = self.anonymous_name(name)
+        context.launch_configurations['anon'][name] = value
+        return value
+
+    def anonymous_name(self, id_value: Text) -> Text:
+        """Get anonymous name based on id value."""
+        import os
+        import random
+        import sys
+        name = f'{id_value}_{os.getpid()}_{random.randint(0, sys.maxsize)}'
```

Comment:
@Blast545 mind to take a look at how ROS 1 used to do this one?

Reply:
The function in master [link to function definition](https://github.com/ros/ros_comm/blob/melodic-devel/tools/rosgraph/src/rosgraph/names.py#L333) Does something similar, I removed the hostname part to avoid any problems related to getting hostname, will add if you think it might be necessary.

Reply:
What kind of problems? As far as I can read, [`socket.gethostname()`](https://docs.python.org/3.8/library/socket.html#socket.gethostname) does not throw.